File: day1/day1.ts
const fs = require('fs');
import { groupbytoobject, map, orderby, pipe, reduce, sum, zip } from 'powerseq';

function parseInput(input: string) {
  return pipe(
    input.split(/\n/),
    map(line => line.trim().split(/\s+/).map(Number)),
    reduce(
      ([accLeft, accRight], [currleft, currRight]) => [accLeft.concat(currleft), accRight.concat(currRight)],
      [[], []] as number[][]
    ),
  );
}

//1.
function part1([left, right]: number[][]) {
  return pipe(
    zip(
      orderby(left, num => num),
      orderby(right, num => num),
      (l, r) => [l, r]
    ),
    sum(([l, r]) => Math.abs(l - r)),
  )
}

//2.
function part2([left, right]: number[][]) {
  const map = groupbytoobject(right, n => n, v => v);
  return sum(left, l => (map[l]?.length || 0) * l);
}

const input = parseInput(fs.readFileSync('./day1/input.txt', 'utf-8'));
console.log(input);
console.log(part1(input));
console.log(part2(input));

File: day1/day1.go
package day1

import (
	"aoc2024/aochelpers"
	"fmt"
	"sort"
	"strings"
)

func parseInput() [][]int {
	file := aochelpers.ReadFileAsString("./day1/input.txt")
	content := strings.Split(string(file), "\n")
	var left, right []int
	for _, line := range content {
		splited := strings.Split(line, "   ")
		a, b := splited[0], splited[1]
		left = append(left, aochelpers.ParseInt(a))
		right = append(right, aochelpers.ParseInt(b))
	}
	return [][]int{left, right}
}

func Part1(){
	parsed := parseInput()
	left, right := parsed[0], parsed[1]
	sort.Ints(left)
	sort.Ints(right)
	var sum int
	for i := 0; i < len(left); i++ {
		sum = sum + aochelpers.AbsInt(left[i] - right[i])
	}
	fmt.Println("Day 1, part 1: ", sum)
}

func Part2(){
	parsed := parseInput()
	left, right := parsed[0], parsed[1]
	
	similarityMap := make(map[int]int)
	for _, num := range right {
		similarityMap[num]++
	}

	var sum int
	for _, num := range left {
		sum += num * similarityMap[num]
	}
	fmt.Println("Day 1, part 2: ", sum)
}
